content/blog/elevate-your-v-project-with-unit-tests.md

```diff
@@ -0,0 +1,749 @@
+a lot of unit tests in vlib that we could use as a reference. This is still valid
+today. Whenever we propose the addition of new code to V's standard library,
+we make sure to include unit tests for it. Additionally, before merging code
+into the main repository, the GitHub Actions CI runner runs a whole bunch of
+tests on it, including all the unit tests. This helps make sure that new code
+does not break any existing code.
+
+This disciplined manner of software development is good to emulate in all V
+projects.
+
+## Prerequisites
+
+We assume that you know your way around the command-line interface. If
+you don't, here a good article to help you get started:
+[The Command Line for Complete Beginners](https://flaviocopes.com/cli-for-beginners/)
+
+Prior exposure to V is required. You can follow this article
+for a quickstart on V: [Getting Started with V](/getting-started-with-v/).
+
+We will pick up from where we left off in the
+[previous article](/the-complete-beginners-guide-to-cli-apps-in-v/),
+so make sure to have your project set up and ready to go.
+
+Alternatively, you can clone the
+[`geo` repository](https://github.com/hungrybluedev/geo), delete the
+`test` directory and follow along with the tutorial. More information
+on V's support for testing can be found in the
+[documentation](https://github.com/vlang/v/blob/master/doc/docs.md#testing).
+
+## A Brief Recap of `geo`
+
+In the [last tutorial](/the-complete-beginners-guide-to-cli-apps-in-v/)
+we created a complete command-line application using the built-in `flag`
+module. So we already have the directory structure like this:
+
+```text
+$ tree .
+.
+├── geometry
+│   ├── metadata.v
+│   ├── options.v
+│   ├── quadrilateral.v
+│   └── triangle.v
+├── geo.v
+└── v.mod
+```
+
+The `v.mod` file anchors the root of the project so that it becomes easier
+to use the `geometry` module. It also contains information about the project
+itself, like the version, description, license, and so on.
+
+The `geo.v` contains the main entry point of the application. It processes
+the command-line arguments and calls the `generate_shape` function with the
+appropriate arguments.
+
+The `geometry` directory contains the source for out module spread across
+multiple files. The `metadata.v` file contains the metadata about the module
+that it reads from the `v.mod`. Next, `options.v` defines the various enums
+and helper structs that help standardise the input. Finally, the `triangle.v`
+and `quadrilateral.v` files contain the actual code for generating the
+desired shapes.
+
+We took an architectural decision to keep the signature of all of the
+shape generation functions same:
+`pub fn generate_xyz(options ShapeOptions) []string`. We will make use
+of this to keep our testing strategy simple.
+
+## First Steps
+
+First, we create a `test` directory. Then we create a `unit_tests` directory.
+Inside that, we create a file `triangle_test.v` in which we will write our
+first test function.
```

Comment:
Why so complicated ?
The _test.v files can go directly inside the root folder. There is no need for a special hierarchy just for them.
Testing should be as simple as possible imho, especially in an introductory blog article about it.

Reply:
It is also an anti-pattern imho to have tests in a separate parallel folder, from the module that they do test.
I really do not understand the reason for the test folder 🤔 .